core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/AvroUpsertMonitor.scala

```diff
@@ -0,0 +1,218 @@
+    }
+}
+
+object AvroUpsertMonitor {
+  case object StartMonitorPass
+
+  def props(
+             pollInterval: FiniteDuration,
+             pollIntervalJitter: FiniteDuration,
+             workspaceService: UserInfo => WorkspaceService,
+             googleServicesDAO: GoogleServicesDAO,
+             samDAO: SamDAO,
+             googleStorage: GoogleStorageService[IO],
+             pubSubDao: GooglePubSubDAO,
+             pubSubSubscriptionName: String,
+             avroUpsertBucketName: String): Props =
+    Props(new AvroUpsertMonitorActor(pollInterval, pollIntervalJitter, workspaceService, googleServicesDAO, samDAO, googleStorage, pubSubDao, pubSubSubscriptionName, avroUpsertBucketName))
+}
+
+class AvroUpsertMonitorActor(
+                                  val pollInterval: FiniteDuration,
+                                  pollIntervalJitter: FiniteDuration,
+                                  workspaceService: UserInfo => WorkspaceService,
+                                  googleServicesDAO: GoogleServicesDAO,
+                                  samDAO: SamDAO,
+                                  googleStorage: GoogleStorageService[IO],
+                                  pubSubDao: GooglePubSubDAO,
+                                  pubSubSubscriptionName: String,
+                                  avroUpsertBucketName: String)
+  extends Actor
+    with LazyLogging
+    with FutureSupport {
+  import AvroUpsertMonitor._
+  import context._
+
+  case class AvroUpsertJson(namespace: String, name: String, userSubjectId: String, userEmail: String, payload: Array[EntityUpdateDefinition])
+
+  implicit val avroUpsertJsonFormat = jsonFormat5(AvroUpsertJson)
+
+  self ! StartMonitorPass
+
+  // fail safe in case this actor is idle too long but not too fast (1 second lower limit)
+  setReceiveTimeout(max((pollInterval + pollIntervalJitter) * 10, 1 second))
+
+  private def max(durations: FiniteDuration*): FiniteDuration = {
+    implicit val finiteDurationIsOrdered = scala.concurrent.duration.FiniteDuration.FiniteDurationIsOrdered
+    durations.max
+  }
+
+  override def receive = {
+    case StartMonitorPass =>
+      // start the process by pulling a message and sending it back to self
+      pubSubDao.pullMessages(pubSubSubscriptionName, 1).map(_.headOption) pipeTo self
+
+    case Some(message: PubSubMessage) =>
+      logger.info(s"received avro upsert message: $message")
+      val (jobId, file) = parseMessage(message)
+      initAvroUpsert(jobId, file).map(response => (response, message.ackId)) pipeTo self
+
+    case None =>
+      // there was no message so wait and try again
+      val nextTime = org.broadinstitute.dsde.workbench.util.addJitter(pollInterval, pollIntervalJitter)
+      system.scheduler.scheduleOnce(nextTime.asInstanceOf[FiniteDuration], self, StartMonitorPass)
+
+    case ((), ackId: String) =>
+      acknowledgeMessage(ackId).map(_ => StartMonitorPass) pipeTo self
+
+    case Status.Failure(t) => throw t
```

Comment:
Again, not fluent in akka... does this catch exceptions thrown by `initAvroUpsert`? Will this kill this actor? Is there a way we can explicitly tell Pub/Sub that this message should be redelivered rather than waiting for ack deadline to expire?

Reply:
I'll partially answer my own question... looks like there is a concept of `nack` but I don't immediately see it in the java client library.